### PR TITLE
[WebProfilerBundle] Display multiple HTTP headers in WDT

### PR DESCRIPTION
**Pull request by ro0NL**
| Q             | A
| ------------- | ---
| Branch?       | 2.7 (partially in 3.2 already)
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no-ish
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | [#20595 (comment)](https://github.com/symfony/symfony/pull/20595#issuecomment-262311651)
| License       | MIT
| Doc PR        | reference to the documentation PR, if any

Backport of https://github.com/symfony/symfony/pull/20595 for 2.7, that includes a fix for displaying multiple HTTP headers in the WDT.

image

//cc @nicolas-grekas this should cover 2.7 - 3.1 right?

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
FYI, I moved the logic in the request collector: https://github.com/symfony/symfony/pull/20595/files#diff-e8f5b14fbfbbeac60fc9f3abe310c3b0R128

**Comment by ro0NL** [comment]
Can we return a `ParameterBag` in 2.7 as well?

**Comment by fabpot** [comment]
Thank you @ro0NL.

File: src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php [hunk 1]
```diff
@@ -40,12 +38,8 @@ public function __construct()
...
     {
         $responseHeaders = $response->headers->all();
-        $cookies = array();
         foreach ($response->headers->getCookies() as $cookie) {
-            $cookies[] = $this->getCookieHeader($cookie->getName(), $cookie->getValue(), $cookie->getExpiresTime(), $cookie->getPath(), $cookie->getDomain(), $cookie->isSecure(), $cookie->isHttpOnly());
-        }
-        if (count($cookies) > 0) {
-            $responseHeaders['Set-Cookie'] = $cookies;
+            $responseHeaders['set-cookie'][] = (string) $cookie;
         }
```
Comment (ro0NL) [inline src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php]: Above should apply to 3.1 & 3.2 as well.

In master we can remove this block due #20569. I covered this with #20567 and should be a separate PR otherwise (basically forgotten in #20569).

File: src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php [hunk 2]
```diff
@@ -121,6 +115,18 @@ public function collect(Request $request, Response $response, \Exception $except
...
+                continue;
+            }
+            if ('request_headers' === $key || 'response_headers' === $key) {
+                $value = array_map(function ($v) { return isset($v[0]) && !isset($v[1]) ? $v[0] : $v; }, $value);
+            }
+            if ('request_server' !== $key && 'request_cookies' !== $key) {
+                $this->data[$key] = $value;
+            }
+        }
+
```
Comment (ro0NL) [inline src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php]: Above should apply to 3.1 as well.

In master this is already applied due #20595.

File: src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php [hunk 3]
```diff
@@ -183,7 +189,7 @@ public function getRequestQuery()
 
     public function getRequestHeaders()
     {
-        return new HeaderBag($this->data['request_headers']);
+        return new ParameterBag($this->data['request_headers']);
```
Comment (ro0NL) [inline src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php]: Same, should be applied to 3.1

File: src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php [hunk 4]
```diff
@@ -203,7 +209,7 @@ public function getRequestAttributes()
 
     public function getResponseHeaders()
     {
-        return new ResponseHeaderBag($this->data['response_headers']);
+        return new ParameterBag($this->data['response_headers']);
```
Comment (ro0NL) [inline src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php]: Same, should be applied to 3.1

File: src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php [hunk 5]
```diff
@@ -302,41 +308,4 @@ public function getName()
...
-        if ($secure) {
-            $cookie .= '; secure';
-        }
-
-        if ($httponly) {
-            $cookie .= '; httponly';
-        }
-
-        return $cookie;
-    }
```
Comment (ro0NL) [inline src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php]: Same, should be applied to 3.1

**Review by ro0NL — COMMENTED** [review]
@fabpot sorry for the inconvenience :) but im not sure upstream merges went as expected..?

edit: is this about 3.1 almost being EOM?
